### PR TITLE
Added parallel cython support for calc grav potential.

### PR DESCRIPTION
**Pull request by pgrete**
## PR Summary

Graviational potential is now calculated using openmp threading.

## PR Checklist

- [x] Code passes flake8 checker
- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

- Remove progress bar in favour of parallel computations.
- Truncation now only works at a thread level as `total_potential` flagged for reduction and thus only accessible after the loop finishes. Please comment on whether this is acceptable or whether it'd be better to introduce a switch to change between parallel and serial calculation)
- Please also comment on the usage of `num_threads`, so that it's in line with general thread parallel coding style in yt

### CODE REVIEW

**Comment by pgrete** [comment]
```python
import numpy as np

import yt
from yt.analysis_modules.level_sets.api import *
from yt.utilities.lib.misc_utilities import gravitational_binding_energy
import time

ds = yt.load("/gfs2/work/nipgrete/IsolatedGalaxy/galaxy0030/galaxy0030")

data_source = ds.disk([0.5, 0.5, 0.5], [0., 0., 1.],
                      (8, 'kpc'), (1, 'kpc'))

def calcPotWithNThreads(num_threads):
    start = time.time()
    print(gravitational_binding_energy(
        data_source[('gas','cell_mass')].in_cgs(),
        data_source[('index','x')].in_cgs(),
        data_source[('index','y')].in_cgs(),
        data_source[('index','z')].in_cgs(),
        False,
        1.,num_threads
    ))
    print("%d thread(s) took %.2f sec" % (num_threads,time.time()-start))

calcPotWithNThreads(1)
calcPotWithNThreads(2)
calcPotWithNThreads(4)
calcPotWithNThreads(8)
calcPotWithNThreads(16)
calcPotWithNThreads(32)

```
Output
```
yt : [INFO     ] 2018-02-07 22:33:21,673 Parameters: current_time              = 0.00600002000283
yt : [INFO     ] 2018-02-07 22:33:21,673 Parameters: domain_dimensions         = [32 32 32]
yt : [INFO     ] 2018-02-07 22:33:21,674 Parameters: domain_left_edge          = [ 0.  0.  0.]
yt : [INFO     ] 2018-02-07 22:33:21,674 Parameters: domain_right_edge         = [ 1.  1.  1.]
yt : [INFO     ] 2018-02-07 22:33:21,674 Parameters: cosmological_simulation   = 0.0
Parsing Hierarchy : 100%|######################################################################################################################################################| 173/173 [00:00<00:00, 2515.67it/s]
yt : [INFO     ] 2018-02-07 22:33:21,760 Gathering a field list (this may take a moment.)
Calculating potential for 100134 cells using 1 threads
3.4658213349e+62
1 thread(s) took 68.52 sec
Calculating potential for 100134 cells using 2 threads
3.4658213349e+62
2 thread(s) took 32.78 sec
Calculating potential for 100134 cells using 4 threads
3.4658213349e+62
4 thread(s) took 16.39 sec
Calculating potential for 100134 cells using 8 threads
3.4658213349e+62
8 thread(s) took 8.20 sec
Calculating potential for 100134 cells using 16 threads
3.4658213349e+62
16 thread(s) took 4.50 sec
Calculating potential for 100134 cells using 32 threads
3.4658213349e+62
32 thread(s) took 2.33 sec
```

File: yt/utilities/lib/misc_utilities.pyx [hunk 1]
```diff
@@ -1021,40 +1024,41 @@ def gravitational_binding_energy(
...
-    total_potential = 0.
+    cdef np.float64_t total_potential = 0.
+    cdef np.float64_t this_potential
 
-    i = 0
     n_q = mass.size
-    pbar = get_pbar("Calculating potential for %d cells" % n_q,
-                    0.5 * (n_q * n_q - n_q))
-    for q_outer in range(n_q - 1):
+    print("Calculating potential for %d cells using %d threads" % (n_q,num_threads))
```
Comment (ngoldbaum) [inline yt/utilities/lib/misc_utilities.pyx]: This should be a log message, not a print statement.

File: yt/utilities/lib/misc_utilities.pyx [hunk 2]
```diff
@@ -1021,40 +1024,41 @@ def gravitational_binding_energy(
...
+
+    # using reversed iterator in order to make use of guided scheduling 
+    # (inner loop is getting more and more expensive)
+    for q_outer in prange(n_q - 1,-1,-1,
+        nogil=True,schedule='guided',num_threads=num_threads):
```
Comment (ngoldbaum) [inline yt/utilities/lib/misc_utilities.pyx]: Depending on how long this takes (e.g. anything more than a few seconds) you probably want to occasionally re-aquire the GIL and check for signals. This means that the calculation will be interruptable by, say, KeyboardInterrupt. Here's an example where we do that elsewhere in some cython code:

https://github.com/yt-project/yt/blob/5e1a06798e2cb694550352d0218e4ecda14e404f/yt/utilities/lib/image_samplers.pyx#L175-L177

It does add some overhead to reaquire the GIL and all threads must stop while the GIL is acquires, so make sure you're not doing it so often that it kills the parallelism but also make sure you're doing it often enough that a user doesn't need to wait more than half a second or so to kill the python interpreter after issuing `KeyboardInterrupt`.

File: yt/utilities/lib/misc_utilities.pyx [hunk 3]
```diff
@@ -1021,40 +1024,41 @@ def gravitational_binding_energy(
...
+            # not using += operator so that variable is not automatically reduced
+            this_potential = this_potential + mass_o * mass_i / \
               sqrt((x_i - x_o) * (x_i - x_o) +
                    (y_i - y_o) * (y_i - y_o) +
                    (z_i - z_o) * (z_i - z_o))
-        i += n_q - q_outer
-        pbar.update(i)
         total_potential += this_potential
-        if truncate and total_potential / kinetic > 1.:
+        if truncate and this_potential / kinetic > 1.:
```
Comment (ngoldbaum) [inline yt/utilities/lib/misc_utilities.pyx]: Why this change?

File: yt/utilities/lib/misc_utilities.pyx [hunk 4]
```diff
@@ -35,6 +35,9 @@ cdef extern from "platform_dep.h":
     # NOTE that size_t might not be int
     void *alloca(int)
 
+from cython.parallel import prange, parallel, threadid
+cimport openmp
```
Comment (ngoldbaum) [inline yt/utilities/lib/misc_utilities.pyx]: `openmp` is never used below

Comment (ngoldbaum) [reply yt/utilities/lib/misc_utilities.pyx]: and in fact, it *has* to be removed, since `cimport openmp` is euivalent to `#include "omp.h"`, which will fail on systems without OpenMP support (e.g. MacOS).

File: yt/utilities/lib/misc_utilities.pyx (hunk 3 above)
Comment (pgrete) [reply yt/utilities/lib/misc_utilities.pyx]: `total_potential` is an (implicitly)  reduced variable. Thus, it can only be accessed after the `prange` loop. Updating it within a `with gil` is also not thread safe as far as I know.

**Comment by ngoldbaum** [comment]
@brittonsmith can you take a look at this?

**Comment by pgrete** [comment]
Here's what the output with the progress bar looks like. `with gil` for interrupt checking for each loop has no measurable effect on performance given the load of the inner loop.
```python
yt : [INFO     ] 2018-02-08 00:12:50,044 Parameters: current_time              = 0.00600002000283
yt : [INFO     ] 2018-02-08 00:12:50,045 Parameters: domain_dimensions         = [32 32 32]
yt : [INFO     ] 2018-02-08 00:12:50,045 Parameters: domain_left_edge          = [ 0.  0.  0.]
yt : [INFO     ] 2018-02-08 00:12:50,045 Parameters: domain_right_edge         = [ 1.  1.  1.]
yt : [INFO     ] 2018-02-08 00:12:50,045 Parameters: cosmological_simulation   = 0.0
Parsing Hierarchy : 100%|######################################################################################################################################################| 173/173 [00:00<00:00, 2517.43it/s]
yt : [INFO     ] 2018-02-08 00:12:50,131 Gathering a field list (this may take a moment.)
Calculating potential for 100134 cells with 1 thread(s): 100%|###########################################################################################################| 100134/100134 [01:05<00:00, 1522.23it/s]
3.4658213349e+62
1 thread(s) took 67.78 sec
Calculating potential for 100134 cells with 2 thread(s): 100%|###########################################################################################################| 100134/100134 [00:33<00:00, 3028.32it/s]
3.4658213349e+62
2 thread(s) took 33.07 sec
Calculating potential for 100134 cells with 4 thread(s): 100%|##########################################################################################################9| 100113/100134 [00:16<00:00, 6029.64it/s]
3.4658213349e+62
4 thread(s) took 16.61 sec
Calculating potential for 100134 cells with 8 thread(s): 100%|#########################################################################################################9| 100095/100134 [00:08<00:00, 12020.17it/s]
3.4658213349e+62
8 thread(s) took 8.33 sec
Calculating potential for 100134 cells with 16 thread(s): 100%|#########################################################################################################6| 99819/100134 [00:04<00:00, 21651.56it/s]
3.4658213349e+62
16 thread(s) took 4.61 sec
Calculating potential for 100134 cells with 32 thread(s):  96%|######################################################################################################    | 96425/100134 [00:02<00:00, 36864.43it/s]
3.4658213349e+62
32 thread(s) took 2.62 sec
```

**Comment by ngoldbaum** [comment]
@yt-fido test this please

**Comment by ngoldbaum** [comment]
OK, tests pass. This needs at least one more reviewer before it can be merged.
